core/src/style/alfonsStyle.cpp

```diff
@@ -0,0 +1,832 @@
+    }
+
+    virtual std::unique_ptr<Mesh> build() override {
+        if (!m_labels.empty()) {
+            m_mesh->setLabels(m_labels, m_scratch.quads);
+        }
+        m_labels.clear();
+        m_scratch.clear();
+
+        return std::move(m_mesh);
+    };
+
+    AlfonsStyle::Parameters applyRule(const DrawRule& _rule, const Properties& _props) const;
+
+    bool prepareLabel(const AlfonsStyle::Parameters& _params, Label::Type _type);
+    void addLabel(const AlfonsStyle::Parameters& _params, Label::Type _type, Label::Transform _transform);
+};
+
+bool Builder::checkRule(const DrawRule& _rule) const {
+    return true;
+}
+
+
+void Builder::addPoint(const Point& _point, const Properties& _props, const DrawRule& _rule) {
+    AlfonsStyle::Parameters params = applyRule(_rule, _props);
+
+    if (!prepareLabel(params, Label::Type::point)) { return; }
+
+    addLabel(params, Label::Type::point, { glm::vec2(_point), glm::vec2(_point) });
+}
+
+void Builder::addLine(const Line& _line, const Properties& _props, const DrawRule& _rule) {
+
+    AlfonsStyle::Parameters params = applyRule(_rule, _props);
+
+    if (!prepareLabel(params, Label::Type::line)) { return; }
+
+    float pixel = 2.0 / (m_tileSize * m_style.pixelScale());
+
+    float minLength = m_scratch.bbox.x * pixel * 0.2;
+
+    for (size_t i = 0; i < _line.size() - 1; i++) {
+        glm::vec2 p1 = glm::vec2(_line[i]);
+        glm::vec2 p2 = glm::vec2(_line[i + 1]);
+        if (glm::length(p1-p2) > minLength) {
+            addLabel(params, Label::Type::line, { p1, p2 });
+        }
+    }
+}
+
+void Builder::addPolygon(const Polygon& _polygon, const Properties& _props, const DrawRule& _rule) {
+    Point p = glm::vec3(centroid(_polygon), 0.0);
+    addPoint(p, _props, _rule);
+}
+
+bool Builder::prepareLabel(const AlfonsStyle::Parameters& _params, Label::Type _type) {
+
+    if (_params.text.empty() || _params.fontSize <= 0.f) {
+        LOGD("invalid params: %s %f", _params.text.c_str(), _params.fontSize);
+        return false;
+    }
+
+    m_scratch.reset();
+
+    m_scratch.fill = _params.fill;
+
+    // Stroke width is normalized by the distance of the SDF spread, then scaled
+    // to a char, then packed into the "alpha" channel of stroke. The .25 scaling
+    // probably has to do with how the SDF is generated, but honestly I'm not sure
+    // what it represents.
+    float fontScale = _params.fontSize / FONT_SIZE * m_style.pixelScale();
+
+    //uint32_t strokeWidth = (_params.strokeWidth / _params.blurSpread * 255. * .25) / fontScale;
+    //m_scratch.stroke = (_params.strokeColor & 0x00ffffff) + (strokeWidth << 24);
+
+    uint32_t strokeWidth = (_params.strokeWidth / 3.f * 255. * .25) / fontScale;
+
+    m_scratch.stroke = (_params.strokeColor & 0x00ffffff) + (strokeWidth << 24);
+
+    {
+        auto ctx = m_style.context();
+
+        std::lock_guard<std::mutex> lock(ctx->m_mutex);
+        auto line = m_shaper.shape(ctx->m_font, _params.text);
```

Comment:
For a first integration using advanced C++ features for the alfons API makes sense. But just for a future note I'd rather not force the user of alfons use C++11/14 features, it's a nice to have within alfons for auto pointers and syntax sugar; but for a more portable usage of the library it would make sense to only expose a more early version of C++ (through headers / API), it might be hard right now because of the number of inclusion files we need for a standard shaping, but we can think of that more if it makes sense to you.